[new locale] Moment.js doesn't have a Chechen (ce) time format support and falls back to Russian, so Notifications are in an incorrect time format

TakhirgeranUmar opened the pull request:
Moment.js doesn't have a Chechen (ce) time format support and falls back to Russian, so Notifications are in an incorrect time format. Please add the file translated into Chechen to "moment/src/locale/".

TakhirgeranUmar committed:
Create ce


[new locale] Moment.js doesn't have a Chechen (ce) time format support and falls back to Russian, so Notifications are in an incorrect time format

TakhirgeranUmar committed:
Rename ce to ce.js

TakhirgeranUmar committed:
Update ce.js

spelling

TakhirgeranUmar commented:
Hello!

I’ve added Chechen language support (locale/ce.js) to Moment.js and would kindly like to ask if this PR could be reviewed.

Chechen is now available in Google Translate, and its digital presence is growing rapidly — especially on Wikipedia, which is one of the fastest-developing Chechen-language platforms.

Adding Chechen to Moment.js would help developers localize applications and improve accessibility for native speakers across educational, cultural, and civic projects.

@ichernev @novemberborn @jmeas — if any of you have a moment to review, I’d be grateful.

If anything in the file needs adjustment, I’m happy to make corrections or improvements.

Thank you for your work and for the opportunity to contribute!

ichernev commented:
You need to follow https://github.com/moment/moment/blob/develop/CONTRIBUTING.md#code-organization and also add tests.

TakhirgeranUmar commented:
@ichernev Thank you for your feedback. I understand the requirements.

    I created the Chechen (ce) locale by adapting the Russian (ru) locale file. I chose the Russian file as a base because I am fluent in Russian, which made the translation process much easier for me. I have translated all the strings (e.g., 'год' → 'шо', 'день' → 'де', 'месяц' → 'бутт') and adjusted the date format order to match the Chechen language convention (Year, Day, Month).

    I need some help with the next steps:

        Tests: I am not a programmer and find the test structure confusing. Could you please point me to the simplest example (perhaps the test file for ru.js?) that I should copy and adapt for ce? I can then translate the test data.

        Code Organization: I reviewed the CONTRIBUTING.md guide. Besides the locale file (src/locale/ce.js), are there any other files I need to update (like a locale index or build configuration)?

    Regarding the translation itself: The linguistic part (all strings and the date format) is correct and follows the rules of the Chechen language. If there is an issue with the code structure or technical implementation, I would be grateful if someone could help fix it. Otherwise, the locale is ready.

    I am committed to making this contribution complete and would greatly appreciate your guidance to finish it correctly. Thank you for your patience and for maintaining this library.

TakhirgeranUmar commented:
@ichernev, thank you for your feedback. I have completed the main part of the work—translating all the strings into the Chechen language. I am not a programmer, and creating the test file is difficult for me. Could you or someone from the team assist with this technical step? I am ready to provide any linguistic clarifications needed for the tests. I will also verify that all Chechen words in the tests are correct.

tacsipacsi commented on ce.js:
```diff
@@ -0,0 +1,208 @@
+//! moment.js локалан конфигураци
+//! locale : Russian [ce]
```
```suggestion
//! locale : Chechen [ce]
```

tacsipacsi replied:
You added _Chechen_, but didn’t remove _Russian_. Please remove line 2.

tacsipacsi replied:
Done, thanks! It looks like I’m not allowed to mark this thread done, even though I started it… So it’ll remain open for now, but anyone with appropriate rights should feel free to close it.

tacsipacsi commented on src/locale/ce.js:
This file should be moved to the `src/locale` directory. I think that’s all what @ichernev means by “following <https://github.com/moment/moment/blob/develop/CONTRIBUTING.md#code-organization>”.

tacsipacsi replied:
This is still not done.

tacsipacsi replied:
It’s now done, but, again, I can’t close the thread.

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

- Added Chechen language support.

- Implemented dual month systems: standalone (Nominative) and format (Locative) for correct date phrasing (e.g., "15 январехь").

- Added comprehensive monthsRegex to support Nominative, Genitive (-ан), and Locative (-ехь) cases.

- Configured pluralization logic for relative time (seconds, minutes, hours, etc.).

TakhirgeranUmar committed:
Rename ce.js to src/locale/ce.js

Moving the locale file to the correct directory as requested by the maintainers.

TakhirgeranUmar commented:
@ichernev @tacsipacsi I have addressed the feedback: the locale file is now in src/locale/ce.jsand I have added comprehensive unit tests insrc/test/locale/ce.js. Please approve the workflow to run the tests. Ready for final review!

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

Changed months to use "январь" instead of "январехь" for better consistency in UI and calendar headers. Updated unit tests accordingly.

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

- Optimized `relativeTime` logic: removed redundant plural forms because in Chechen, nouns do not change their form after cardinal numbers in this context.
- Updated `calendar` strings: changed "Next/Last" prefixes to modern "РогӀера/Хьалхара" for better UI consistency.
- Refined `monthsShort` and `monthsRegex` to improve parsing accuracy.
- Cleaned up the code by removing redundant switch-cases and duplicated functions.
- Added English comments for better maintainability.

tacsipacsi commented:
Thanks for renaming the file! However, I don’t see the tests you mentioned – did you forget to `git add` the file (in case you use the command-line Git tools)?

P.S. Myself being new to this project, I’m not able to approve the workflow. I hope @ichernev will be able to.

TakhirgeranUmar commented:
@tacsipacsi. Thank you so much for your help! Here is the link:
https://github.com/TakhirgeranUmar/moment/blob/master/src/test/locale/ce.js

tacsipacsi commented:
I see. You have to put it on the `patch-2` branch (see the header of the current page: _@TakhirgeranUmar wants to merge 12 commits into [moment:develop](https://github.com/moment/moment/tree/develop) **from [TakhirgeranUmar:patch-2](https://github.com/TakhirgeranUmar/moment/tree/patch-2)**_), not on the `master` branch. That is, the following link should not be a 404: https://github.com/TakhirgeranUmar/moment/blob/patch-2/src/test/locale/ce.js. (If you’re using the GitHub web interface, I think you should be able to create the file using the following link: https://github.com/TakhirgeranUmar/moment/new/patch-2/src/test/locale. I don’t have write access to your repo, so I can’t do it for you.)

TakhirgeranUmar committed:
Create ce.js

TakhirgeranUmar commented:
I have a question about the relativeTime strings. Could you please clarify where exactly the future and past strings are displayed in the user interface?

The reason I ask is that for the future prefix, it might be more natural in Chechen to use words like 'йисина', 'дисина', or 'бисина' (which mean 'remaining'), depending on the noun's grammatical class (minutes, hours, or months).

If these strings are often used for countdowns or status indicators, I would like to make sure the translation feels as natural as possible. Knowing the context will help me decide if I should stick with the current version or implement a more specific one.

tacsipacsi commented:
Moment.js is a library, used by many different projects for many different purposes. There is no “the user interface”, and I don’t think there is any chance telling what these translations will mainly be used for.

TakhirgeranUmar committed:
Update ce.js

Refactor relativeTime to use literary grammatical classes (J, D, B)

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar commented:
Updated the locale and tests.

I have implemented the following changes:

Grammatical Classes: Refactored relativeTime to use proper literary grammatical classes for future tense:

Й (J): йаьлча (seconds, minutes)

Д (D): даьлча (hours, days, weeks, years)

Б (B): баьлча (months)

Code Quality: Translated all internal comments to English to meet the project's standards.

Tests: Updated src/test/locale/ce.js with comprehensive tests covering all grammatical classes and time units (including seconds).

Ready for another review.

TakhirgeranUmar committed:
Update ce.js

Grammatical Classes: Implemented dynamic handling of Chechen noun classes (J, D, B) for relative time. Future suffixes (йаьлча, даьлча, баьлча) are applied based on the time unit's gender.

Calendar System: * Set Monday as the first day of the week (dow: 1).

Added full and abbreviated names for months and weekdays in the Nominative case.

Time of Day: Integrated meridiem support to reflect Chechen periods of the day: буьйса (night), Ӏуьйре (morning), де (day), and суьйре (evening).

Regex Support: Robust regular expressions for parsing dates in various cases (Nominative, Genitive, Locative).

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

Added Chechen locale with support for specific grammatical gender classes (J, D, B) in relative time formatting.

Implemented dynamic suffixes for future time (йаьлча, даьлча, баьлча) and correct word order for past time (%s хьалха).

Set Year.Day.Month (YYYY.DD.MM) as the standard numeric date format.

Added comprehensive regex for month parsing to support various case inflections (nominative, genitive, and locative).

Configured day periods (meridiem) and week settings (Monday start).

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

tacsipacsi commented on src/locale/ce.js:
```diff
@@ -0,0 +1,121 @@
+}
+
+var monthsParse = [
+    /^янв/i, /^фев/i, /^мар/i, /^апр/i, /^ма[й]/i, /^июн/i,
```
```suggestion
    /^янв/i, /^фев/i, /^мар/i, /^апр/i, /^май/i, /^июн/i,
```
In Russian, the [character class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions#atoms) in `/^ма[йя]/i` used written to accept both _май_ and _мая_. If there is only one version in Chechen, the character class (brackets) is not necessary.

TakhirgeranUmar replied:
I have updated the monthsParse array with the following changes:

May: Removed the unnecessary character class (/^май/i). In Chechen, the stem is stable and doesn't change its ending (like the Russian 'май/мая'), so the brackets were redundant.

November: Updated the pattern to /^нояб/i. In Chechen, the root for November is 'Ноябрь' and its inflected forms (e.g., 'Ноябран', 'Ноябрехь') retain the 'б'. This also aligns the parser with the monthsShort array already defined in the locale.

These changes ensure better linguistic accuracy and consistency within the configuration.

Thank you for your feedback! If you have any further suggestions or remarks, I am ready to make additional improvements.

tacsipacsi commented on src/locale/ce.js:
```diff
@@ -0,0 +1,121 @@
+    monthsShort: 'янв._февр._мар._апр._май_июнь_июль_авг._сент._окт._нояб._дек.'.split('_'),
+    weekdays: 'кӀиранде_оршот_шинара_кхаара_йеара_пӀераска_шот'.split('_'),
+    weekdaysShort: 'кӀир_орш_шин_кхар_йеа_пӀер_шот'.split('_'),
+    weekdaysMin: 'кӀир_орш_шин_кхар_йеа_пӀер_шот'.split('_'),
```
Isn’t there an even more compact way to write week days? [`weekDaysMin` should be used when there is very little space available](https://momentjs.com/docs/#/customization/weekday-min/). In English, `weekdaysShort` is _Sun_, _Mon_ etc., while `weekdaysMin` is _Su_, _Mo_ etc. – just enough characters to allow users to tell day names apart. (However, if there isn’t a common way to write weekdays that compact in Chechen, making up your own abbreviations is probably not a good idea.)

TakhirgeranUmar replied:
@tacsipacsi The weekdaysMin is currently kept at 3–4 characters to ensure readability. In Chechen, many sounds are represented by digraphs (e.g., кӀ, пӀ, кх). These are treated as single letters in the alphabet.

If we were to use a strict 2-character limit (like in English), it would break these digraphs (e.g., 'ПӀ' becoming 'П'), which is linguistically incorrect and changes the sound/letter entirely.

It is also worth noting that there are no official calendars or established linguistic standards for two-letter abbreviations in Chechen. Creating them arbitrarily could lead to confusion for native speakers. However, if there is a strong preference for a more compact version using full phonemes (e.g., 'ПӀ', 'Кх', 'Ор'), I am open to updating the code. For now, the 3-character version is the most reliable.

tacsipacsi replied:
You should absolutely not break digraphs (unless it’s common in Chechen; in German, for example, AFAIK it happens that the trigraph _sch_ gets broken in such situations), nor make up anything that won’t be understood by Chechen speakers.

(My native language [Hungarian](https://github.com/moment/moment/blob/develop/src/locale/hu.js) has even shorter weekday names on average than English, being usually only one character long, except for the two-character digraph _Cs_ and Wednesday and Saturday, which both begin with the same digraph _Sz_, so a second letter / third character is also needed. However, these abbreviations are common and well-understood in Hungarian, for example my 2025 wall calendar also includes them. If Chechen doesn’t have such short abbreviations, then don’t use them.)

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar committed:
Update ce.js

TakhirgeranUmar commented:
@ichernev Hello! Just a friendly reminder about this PR for the Chechen (ce) locale. It has been ready and passing all checks for quite some time now (over 3 months). It would be great to see this merged so we can finally use the native localization. Could you please take a look when you have a chance? Thank you!

mattjohnsonpint reviewed (changes requested):
Thank you for adding Chechen support and for continuing to improve this PR based on the earlier feedback. I’m reviewing this with AI assistance because I’m not a native Chechen speaker, so I’m relying on published language data and would appreciate your help confirming the linguistic details.

I reviewed the changes after rebasing them locally onto the latest `develop` and found a few things that need addressing before they can be merged:

1. The calendar test currently fails five assertions. It creates a fixed reference date but calls `.calendar()` without passing that reference date. For example:

```js
moment(now).subtract({ d: 1 }).calendar(now)
```

Also, dates exactly seven days away use Moment’s `sameElse` format rather than `lastWeek` or `nextWeek`, so those cases should use a difference of two to six days and explicitly cross a week boundary.

2. The locale and test files currently fail ESLint’s `one-var` rule and the project’s Prettier check. Running `pnpm validate` stops at these four lint errors.

3. CLDR uses `яьлча`, rather than `йаьлча`, for future minutes and seconds:

- Minute: `{0} минот яьлча`
- Second: `{0} секунд яьлча`

Source: [Unicode CLDR 48.2 Chechen relative-time data](https://raw.githubusercontent.com/unicode-org/cldr-json/48.2.1/cldr-json/cldr-dates-full/main/ce/dateFields.json)

Could you confirm whether `йаьлча` is an intentional contemporary spelling? If there is no authoritative source supporting it, I think the locale and tests should use `яьлча`.

4. Several inflected month names listed in the regular expressions cannot currently be parsed strictly. For example, all of these return an invalid Moment:

```js
moment('мартан', 'MMMM', 'ce', true)
moment('мартехь', 'MMMM', 'ce', true)
moment('майхь', 'MMMM', 'ce', true)
moment('августан', 'MMMM', 'ce', true)
```

This happens because shorter alternatives such as `март` and `август` occur before the longer forms in the regular expression. The longer alternatives need to come first, and tests should cover every supported form.

5. The relative-time function ignores Moment’s `withoutSuffix` argument. As a result, a neutral duration still includes future wording:

```js
moment.duration(1, 'day').locale('ce').humanize()
// "де даьлча"
```

I would expect the suffixless result to be `де`, while `humanize(true)` would produce `де даьлча`. Please confirm whether that distinction is grammatically appropriate in Chechen.

6. Uppercase day periods are accepted by `meridiemParse`, but `isPM` handles only lowercase input:

```js
moment('3 де', 'h A', 'ce', true).hour() // 15
moment('3 ДЕ', 'h A', 'ce', true).hour() // 3
```

Making the `isPM` check case-insensitive should make parsing consistent.

I also found some locale-data differences that need either adjustment or supporting sources:

- CLDR abbreviates the months as `янв`, `фев`, `мар`, `апр`, `май`, `июн`, `июл`, `авг`, `сен`, `окт`, `ноя`, and `дек`. The PR uses several different lengths and adds punctuation.
- CLDR supplies shorter weekday forms than the PR.
- I could not find an authoritative source for the numeric `YYYY.DD.MM` year-day-month order.
- I could not find a Chechen source defining the proposed night, morning, day, and evening boundaries.

Source: [Unicode CLDR 48.2 Chechen locale data](https://raw.githubusercontent.com/unicode-org/cldr/release-48-2/common/main/ce.xml)

I do not think the full weekday names should simply be changed to match CLDR. CLDR lists `кӀира`, `еара`, and `шуот`, but the Nichols-Vagapov *Chechen-English and English-Chechen Dictionary* supports the PR’s `кӀиранде`, `йеара`, and `шот`:

- [Sunday](https://books.google.com/books?vid=ISBN0415315948&jscmd=SearchWithinVolume2&q=Sunday)
- [Thursday](https://books.google.com/books?vid=ISBN0415315948&jscmd=SearchWithinVolume2&q=Thursday)
- [Saturday](https://books.google.com/books?vid=ISBN0415315948&jscmd=SearchWithinVolume2&q=Shot)
- [Dictionary project background](https://linguistics.berkeley.edu/~chechen/chdictionaries.html)

That appears to be a genuine source conflict, and the dictionary provides reasonable support for your full weekday names. Contemporary institutional or published examples would make the case even stronger.

Finally, could you switch the tests to the project’s standard `localeModule('ce')` helper and add coverage for all months, weekdays, relative-time units, suffixless output, strict parsing, and meridiem boundaries? That helper automatically runs the common locale tests and restores the previous locale after each test.

I appreciate the linguistic work you have already put into this. Once the failing tests and validation errors are fixed, and the remaining differences are either aligned with CLDR or supported with reliable sources, I’ll be happy to review the update.

mattjohnsonpint commented:
Note, I have hidden all previous comments. Not that there was anything wrong, but a lot of time has passed. Let's focus on my review and changes you can make to address the points given.  Please rebase your PR on the `develop` branch also.  Thanks.
